cli/azd/cmd/deploy.go

```diff
@@ -142,38 +149,37 @@ func (d *deployAction) Run(ctx context.Context) (*actions.ActionResult, error) {
 		Title: "Deploying services (azd deploy)",
 	})
 
-	var svcDeploymentResult project.ServiceDeploymentResult
-	var deploymentResults []project.ServiceDeploymentResult
+	var svcDeploymentResult *project.ServiceDeployResult
+	var deploymentResults []*project.ServiceDeployResult
 
-	for _, svc := range proj.Services {
+	for _, svc := range d.projectConfig.Services {
 		// Skip this service if both cases are true:
 		// 1. The user specified a service name
 		// 2. This service is not the one the user specified
-		if d.flags.serviceName != "" && svc.Config.Name != d.flags.serviceName {
+		if d.flags.serviceName != "" && svc.Name != d.flags.serviceName {
 			continue
 		}
 
-		stepMessage := fmt.Sprintf("Deploying service %s", svc.Config.Name)
+		stepMessage := fmt.Sprintf("Deploying service %s", svc.Name)
 		d.console.ShowSpinner(ctx, stepMessage, input.Step)
-		result, progress := svc.Deploy(ctx, d.azdCtx)
 
-		// Report any progress to logs only. Changes for the console are managed by the console object.
-		// This routine is required to drain all the string messages sent by the `progress`.
+		deployTask := d.serviceManager.Deploy(ctx, svc)
+
 		go func() {
-			for message := range progress {
-				log.Printf("- %s...", message)
+			for progress := range deployTask.Progress() {
+				updatedMessage := fmt.Sprintf("Deploying service %s (%s)", svc.Name, progress.Message)
+				d.console.ShowSpinner(ctx, updatedMessage, input.Step)
```

Comment:
I really wish we could get rid of this reporting-strategy. It limits what it can be done by using the `console` object directly.

Using `string` only as the `progess.Message` makes it impossible for anyone getting the Message to know anything else about the message. 

In this case, for example, we are assuming that all messages are only meant to be part of the spinner and we are using a hardcoded message as prefix to keep the spinner running. Changing only the description at the end.

What happens if we need to send a different message. Like a warning message, or if we want to persist a message on top of the spinner, or if we want to stop the spinner and interact with I/O.  Or if we want to indent the messages and add a bullet before the message.  Any of these things should be easy to find and use from one single approach/strategy.

The `console` is already an `interface` which enable us to create the implementation for any time of message and for as many outputs as we want (a terminal, json, or any other output).  We don't need to have an extra layer/component to define how components use the `console`.

Reply:
I agree with you that we need a refactor on the UX side to deal with how data is passed from the data tier to the UX but this PR is not that.  

Nothing changes here from before except we now get that incremental progress state during deployment so the user is able to visual what step the deployment is on such as `Building Container Image`, `Compressing package artifacts`, etc.

Reply:
The struct that is returned from the commands is extensible and can return anything that we want.  If we want to add more structured data, level of verbosity, etc... we can and only need to change the shape of our `ServiceProgress` struct.

Reply:
> Nothing changes here from before

We are making possible to use the `progress` to update the UX view.  So there will be two ways where the UX output is produced while doing `provision`.